src/TSTransformer/nodes/class/transformDecorators.ts

```
@@ -0,0 +1,289 @@
+	node: ts.ClassLikeDeclaration,
+	classId: luau.AnyIdentifier,
+): luau.List<luau.Statement> {
+	const result = luau.list.make<luau.Statement>();
+	const finalizers = luau.list.make<luau.Statement>();
+
+	const multipleDecorators = node.decorators !== undefined && node.decorators.length > 1;
+
+	for (const decorator of node.decorators ?? []) {
+		// eslint-disable-next-line no-autofix/prefer-const
+		let [expression, prereqs] = state.capture(() => transformExpression(state, decorator.expression));
+
+		luau.list.pushList(result, prereqs);
+
+		if (multipleDecorators) {
+			const tempId = luau.tempId("decorator");
+			luau.list.push(
+				result,
+				luau.create(luau.SyntaxKind.VariableDeclaration, {
+					left: tempId,
+					right: expression,
+				}),
+			);
+			expression = tempId;
+		}
+
+		assert(luau.isIndexableExpression(expression));
+
+		// Class = decorator(Class) or Class
+		luau.list.unshift(
+			finalizers,
+			luau.create(luau.SyntaxKind.Assignment, {
+				left: classId,
+				operator: "=",
+				right: luau.binary(luau.call(expression, [classId]), "or", classId),
+			}),
+		);
+	}
+
+	luau.list.pushList(result, finalizers);
+
+	return result;
+}
+
+export function transformDecorators(
+	state: TransformState,
+	node: ts.ClassLikeDeclaration,
+	classId: luau.AnyIdentifier,
+): luau.List<luau.Statement> {
+	const result = luau.list.make<luau.Statement>();
+
+	// https://www.typescriptlang.org/docs/handbook/decorators.html#decorator-evaluation
+
+	// Instance Decorators
+	for (const member of node.members) {
+		if (!ts.getSelectedSyntacticModifierFlags(member, ts.ModifierFlags.Static)) {
+			if (ts.isMethodDeclaration(member)) {
+				luau.list.pushList(result, transformMethodDecorators(state, member, classId));
+				luau.list.pushList(result, transformParameterDecorators(state, member, classId));
+			} else if (ts.isPropertyDeclaration(member)) {
+				luau.list.pushList(result, transformPropertyDecorators(state, member, classId));
+			}
```

Silent fail on unknown case if anything is ever added, should probably assert false

---

Same below on lines 269-274